kong/cmd/utils/process_secrets.lua

```diff
@@ -0,0 +1,195 @@
+local b64 = require "ngx.base64"
+local cjson = require "cjson.safe"
+local file = require "pl.file"
+local path = require "pl.path"
+local cipher = require "resty.openssl.cipher"
+local digest = require "resty.openssl.digest"
+local rand = require "resty.openssl.rand"
+
+
+local fmt = string.format
+local sub = string.sub
+local type = type
+local pairs = pairs
+
+
+local CIPHER_ALG = "aes-256-gcm"
+local DIGEST_ALG = "sha256"
+local IV_SIZE = 12
+local TAG_SIZE = 16
+local AAD = fmt("%s|%s", CIPHER_ALG, DIGEST_ALG)
+
+
+local function read_key_data(key_data_path)
+  if not path.exists(key_data_path) then
+    return nil, fmt("failed to read key data (%s): file not found", key_data_path)
+  end
+
+  local key_data, err = file.read(key_data_path, true)
+  if not key_data then
+    return nil, fmt("failed to read key data file: %s", err)
+  end
+
+  return key_data
+end
+
+
+local function hash_key_data(key_data)
+  local hash, err = digest.new(DIGEST_ALG)
+  if not hash then
+    return nil, fmt("unable to initialize digest (%s)", err)
+  end
+
+  local ok
+  ok, err = hash:update(key_data)
+  if not ok then
+    return nil, fmt("unable to update digest (%s)", err)
+  end
+
+  local key
+  key, err = hash:final()
+  if not key then
+    return nil, fmt("unable to create digest (%s)", err)
+  end
+
+  return key
+end
+
+
+local function extract(conf)
+  local refs = conf["$refs"]
+  if not refs or type(refs) ~= "table" then
+    return
+  end
+
+  local secrets = {}
+  for k in pairs(refs) do
+    secrets[k] = conf[k]
+  end
+
+  return secrets
+end
+
+
+local function encrypt(plaintext, key_data)
```

Comment:
It feels that we have this type of code (at least _very_ similar) in quite some places in Kong. We should think about a separate module for that _sometime_ in the future.

Reply:
@jschmid1, I know, but I feel we don't have anything like that in `CE`, only in `EE`.